### PR TITLE
Data Deletion Request Handing spec

### PR DESCRIPTION
**Pull request by alextcone**
For CCPA/US Privacy Technical Working Group approval before merge

### CODE REVIEW

File: CCPA/Data Deletion Request Handling.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
+# Data Deletion Request Handling
+
+## Summary
+
+This specification provides a way for a publisher to communicate to downstream parties via a webpage about a consumer's request to delete their data. 
+
+In particular, Section 1798.105(c) of the California Consumer Protection Act (CCPA) states that "[a] business that receives a verifiable consumer request from a consumer to delete the consumer's personal information [shall] . . direct any service providers to delete the consumer's personal information from their records." 
+
+In response to this directive in the CCPA, this specification offers the technical means for publishers to communicate deletion requests to their service providers. All service providers covered by IAB Privacy LLC's Limited Service Provider Agreement (LSPA) and any other providers who receive the signal for deletion requests are then enabled and obliged to comply and delete the specified consumer data.
```
Comment (alextcone) [inline CCPA/Data Deletion Request Handling.md:9]: remove "and obliged" (comment from TL counsel)

File: CCPA/Data Deletion Request Handling.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+
+
+### Multiple Services
+
+For vendors that provide multiple services, a URL is provided for each service. A publisher may use any of those services but not necessarily all of them. When a consumer requests data deletion, only the scripts provided on the page will trigger the delete request back to the vendor. Vendors can then delete the personal data, in accordance with their practices, for each service provided on the publisher property where the request originated.
+
+
+### Deletion
+
+The technical solution detailed in this specification provides the means to signal consumer requests for data deletion. Companies supporting the US Privacy Framework will respond to the signals by deleting the consumer's relevant personal data. The process for deletion depends on the company's technology and operational practices in place. _How_ a vendor deletes a consumer's personal data is out of scope for this specification.
```
Comment (alextcone) [inline CCPA/Data Deletion Request Handling.md:33]: Change to read:

> The technical solution detailed in this specification provides the means to signal consumer requests for data deletion. Companies supporting the US Privacy Framework **(i.e., service providers)** will respond to the signals by deleting the consumer's relevant personal data to the extent required by CCPA. The process for deletion depends on the company's technology and operational practices in place. How a vendor deletes a consumer's personal data is out of scope for this specification.

bolded text is new

File: CCPA/Data Deletion Request Handling.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+[Limited Service Provider Agreement](https://www.iabprivacy.com/lspa-2019-12.pdf) (12-2-19 version)
+
+
+## How it Works
+
+Any ad tech company (vendor) supporting the US Privacy framework, exposes a JavaScript resource for each service they offer. The vendor hosts a URL that provides this resource and also serves as a unique identifier for the resource. The URL is made public to a maintained LSPA signatories list and to non-signatories based on the vendor's documented practices.
+
+Publishers working with any services provided by these vendors consume the exposed JavaScript and embed it on their properties where consumers can request deletion of their personal data. If data deletion is requested, all javascript signals on the publisher property are sent to the associated vendors. Vendors then respond by deleting the relevant personal data and signaling any affected vendors downstream.
+
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:24]: "...and signaling any affected vendors downstream."
I thought we were agreed that this was a one-level delete and not some kind of recursive fan out. Consider removing?

File: CCPA/Data Deletion Request Handling.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+### Non-web Environments
+
+If the consumer initiates a data deletion request from a non-web environment, the request can be propagated through the same javascript resource described above by directing the user to a publisher webpage to complete the deletion request. The request can be associated to the device from where it originated using defined parameters.
+
+
+## Deletion Signaling
+
+Deletion signaling uses two commands: `registerDeletion` and `performDeletion`. The first command, `registerDeletion`, is executed when the publisher loads script on a given page or property. This informs the vendor that the provided script has been loaded. The second command, `performDeletion`, is executed if a consumer requests data deletion and informs the vendor that a consumer has requested that their private data be deleted.
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:43]:
```suggestion
Deletion signaling uses two commands: `registerDeletion` and `performDeletion`. The first command, `registerDeletion`, is executed when the publisher loads a vendor script on a given page or property. This registers the vendor script to receive the signal sent by the `performDeletion` call. The second command, `performDeletion`, is executed if a consumer requests data deletion and signals the vendor that a consumer has requested that their personal data be deleted.  
```

Not sure about the first sentence change but I am fairly sure that the call to registerDeletion does not signal the vendor in any way. Or at least, does not _need_ to signal anything to the vendor.

File: CCPA/Data Deletion Request Handling.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+This command registers a vendor-specific callback function at the API. When the publisher loads the vendor deletion script, it must execute `registerDeletion`. 
+
+The `registerDeletion` command uses the following syntax: 
+
+```
+__usapi("registerDeletion", version, performDeletionFunction)
+```
+
+Upon execution, the vendor is notified that the vendor's deletion script was loaded.
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:72]: Not sure this is true; see my earlier comment. Maybe rephrase to:
```suggestion
Upon execution, the vendor script is registered to receive the signal sent by the call to `performDeletion`.
```

File: CCPA/Data Deletion Request Handling.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+
+## performDeletion
+
+The publisher, or its CMP where applicable, calls this command on some user delete action to initiate the deletion process. The command communicates to vendors that a specified user has requested that their personal data be deleted.
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:77]:
```suggestion
The publisher, or its CMP where applicable, calls this command based on some user delete action to initiate the deletion process. The command communicates to vendors that a specified user has requested that their personal data be deleted by calling the `performDeletionFunction` registered during the call to `registerDeletion`. 
```

File: CCPA/Data Deletion Request Handling.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+*   If the iframe sandbox parameter is provided, the following restrictions are lifted:
+    *   `allow-scripts`
+    *   `allow-storage-access-by-user-activation`
+
+
+## registerDeletion
+
+This command registers a vendor-specific callback function at the API. When the publisher loads the vendor deletion script, it must execute `registerDeletion`. 
+
+The `registerDeletion` command uses the following syntax:
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:66]: Change in line with the performDeletion text below.
```suggestion
This command is an update to the existing `__uspapi` function with the following syntax: 
```

File: CCPA/Data Deletion Request Handling.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+```
+
+Upon execution, the vendor is notified that the vendor's deletion script was loaded.
+
+
+## performDeletion
+
+The publisher, or its CMP where applicable, calls this command on some user delete action to initiate the deletion process. The command communicates to vendors that a specified user has requested that their personal data be deleted. 
+
+This command is an update to existing  `__uspapi`. The `performDeletion` command uses the following syntax:
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:79]:
```suggestion
This command is an update to the existing `__uspapi` function with the following syntax: 
```

File: CCPA/Data Deletion Request Handling.md [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+        var json = msgIsString ? JSON.parse(event.data) : event.data;
+       }
+       catch(e)
+       {
+        var json = null;
+       }
+       if(typeof (json) === 'object' && json !== null && '__uspapiCall' in json)
+       {
+        var i = json.__uspapiCall;
+        window.__uspapi(i.command, i.version, function (retValue, success)
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:219]: Does this need the extra optional parameter ('identifiers') used in non-web contexts as a placeholder?

File: CCPA/Data Deletion Request Handling.md [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    break;
+  }
+  f = f.parent;
+}
+
+/* Set up a __uspapi function to do the postMessage and
+stash the callback.
+This function behaves (from the caller's perspective)
+identically to the in-frame __uspapi call */
+window.__uspapi = function(cmd, ver, callback, param) {
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md:290]: This has always somewhat confused me, mainly because I'm not a JS expert (more of a novice). Does this not overwrite the actual `__uspapi` function with this one?

File: CCPA/Data Deletion Request Handling.md (hunk 4 above)
Comment (Facens) [reply CCPA/Data Deletion Request Handling.md:43]: @lon-pilot-mpf thanks for the edits! Proposing some further improvements for readability:

"is executed when the publisher loads a vendor script on a given page or property"
->
"is executed by the vendor script when the publisher loads it on a given page or property"

"The second command, `performDeletion`, is executed"
->
"The second command, `performDeletion`, is executed by the publisher"

File: CCPA/Data Deletion Request Handling.md [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+Vendors need access to the publisher domain to get a user identifier to associate with a delete request. In order for vendors to do so, the vendor scripts described in this specification must be loaded _directly _on the publisher domain normally used for collecting data and serving ads and not on a separate or generic domain. For safety, the vendor scripts can be loaded in an iframe, provided the following conditions are met:
+*   The iframe is hosted on the publisher domain
+*   If the iframe sandbox parameter is provided, the following restrictions are lifted:
+    *   `allow-scripts`
+    *   `allow-storage-access-by-user-activation`
+
+
+## registerDeletion
+
+This command registers a vendor-specific callback function at the API. When the publisher loads the vendor deletion script, it must execute `registerDeletion`.
```
Comment (Facens) [inline CCPA/Data Deletion Request Handling.md:64]:
```suggestion
This command registers a vendor-specific callback function at the API. When the publisher loads the vendor deletion script, those scripts must execute `registerDeletion`. 
```

**Comment by alextcone** [comment]
@Facens and @lon-pilot-mpf there are two more questions on the code example that I myself cannot resolve. Ideally we could publish this this week (as I'm already a week behind my artificial deadline). How do you suggest we resolve? If we did resolve those questions would you both give an official approval? Would like to have two from the group here!

**Comment by Facens** [comment]
@alextcone I don't have the competence to respond. I think you should loop in Jan or someone else who's savvy on the JS side.

File: CCPA/Data Deletion Request Handling.md (hunk 9 above)
Comment (janwinkler) [reply CCPA/Data Deletion Request Handling.md:219]: yes, probably

File: CCPA/Data Deletion Request Handling.md (hunk 10 above)
Comment (janwinkler) [reply CCPA/Data Deletion Request Handling.md:290]: yes it does. but it should only be called/executed if a __usapi function is NOT found in the page. it then serves as a proxy for communicating between iframes

Comment (janwinkler) [reply CCPA/Data Deletion Request Handling.md:290]: not sure the brackets are set correct in the example

File: CCPA/Data Deletion Request Handling.md (hunk 9 above)
Comment (alextcone) [reply CCPA/Data Deletion Request Handling.md:219]: can you show exactly where that would go?

File: CCPA/Data Deletion Request Handling.md [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+    </script>
+  </body>
+</html>
+```
+
+Vendor script (at `https://vendorx.com/ccpa-delete-function.js`)
+
+```
+// find the __uspapi frame
+if(!('__uspapi' in window)){
```
Comment (alextcone) [inline CCPA/Data Deletion Request Handling.md:271]: @janwinkler did I do this correctly?

File: CCPA/Data Deletion Request Handling.md [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+
+
+## How it Works
+
+Any ad tech company (vendor) supporting the US Privacy framework, exposes a JavaScript resource for each service they offer. The vendor hosts a URL that provides this resource and also serves as a unique identifier for the resource. The URL is made public to a maintained LSPA signatories list and to non-signatories via the vendor's method of choice.
+
+Publishers working with any services provided by these vendors consume the exposed JavaScript and embed it on their properties where consumers can request deletion of their personal data. If data deletion is requested, all vendor javascript on the publisher property is notified. Vendors then respond by deleting the relevant personal data and signaling any affected vendors downstream.
+
+
+### Multiple Services
```
Comment (chrispaterson) [inline CCPA/Data Deletion Request Handling.md:26]: From the blurb above in the CCPA Regulation:

> direct any service providers to delete the consumer's personal information from their records

The relationship is consumer to service **_provider_** not consumer to **_service_**.  So, therefore, no matter how many services a Vendor provides, there is only one request to be made to that Vendor and therefore no need to support multiple services or to even specify that in this technical document.

File: CCPA/Data Deletion Request Handling.md [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+
+For vendors that provide multiple services, a URL is provided for each service. A publisher may use any of those services but not necessarily all of them. When a consumer requests data deletion, only the scripts provided on the page will trigger the delete request back to the vendor. Vendors can then delete the personal data, in accordance with their practices, for each service provided on the publisher property where the request originated.
+
+
+### Deletion
+
+The technical solution detailed in this specification provides the means to signal consumer requests for data deletion. Companies supporting the US Privacy Framework (i.e., service providers) will respond to the signals by deleting the consumer's relevant personal data to the extent required by CCPA. The process for deletion depends on the company's technology and operational practices in place. _How_ a vendor deletes a consumer's personal data is out of scope for this specification.
+
+
+### Non-web Environments
```
Comment (chrispaterson) [inline CCPA/Data Deletion Request Handling.md:36]: "In-App" is correct term here.  "In-App" was created specifically for this use case.  I understand it has a Mobile-only feel but that's only because Mobile is the biggest use case.  If a clarifying statement is needed then add that.

My problem with Non-web is that it then proceeds to tell us to use a web page in the non-web environment.  It sounds like there is no Internet connection or something.

File: CCPA/Data Deletion Request Handling.md [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+```
+
+The callback parameter of the `__uspapi` is not used in this case and can remain null. The `identifiers` parameter is only required when handling non-web deletion requests and is further explained in the following section. 
+
+
+### Identifying Data to Delete in Non-web Environments
+
+When operating in a non-web environment, data deletion requests are handled by sending the user to a web page where they can complete the request to have their data deleted. Vendors need certain information to correctly identify the data to delete:  the platform name, the unique app identifier used in the app store, and the device identifier for that platform / store. 
+
+The `performDeletion` command includes a parameter for identifiers. Using the `identifiers` parameter, publishers can pass multiple items, each with the required fields: `platform`, `app_identifier`, and `user_identifier`. These details should be passed from the app context to the delete webpage and then along with the request from the page. Below are a few examples of that information:
```
Comment (chrispaterson) [inline CCPA/Data Deletion Request Handling.md:92]: I don't think it is ever possible to have more than one `platform`, `app_identifier`, or `user_identifier` in a single app.

File: CCPA/Data Deletion Request Handling.md [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+{
+  "platform": "android-play-store",
+  "app_identifier": "com.acmeinc.acmeapp",
+  "user_identifier": "aaaacdda123802ae-11fb191c-1247abad-12340794ad123394ac123912"
+...
+}
+```
+
+```
+Note: For purposes of this explanation, a "bundle id" is used for (iOS) and a "package" for (Android).
```
Comment (chrispaterson) [inline CCPA/Data Deletion Request Handling.md:111]: for which property?

File: CCPA/Data Deletion Request Handling.md [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+```
+
+### Common Platform / Stores identifiers
+
+These identifiers can be used for the "platform" field where applicable. Otherwise, the platform field can include an arbitrary value.
```
Comment (chrispaterson) [inline CCPA/Data Deletion Request Handling.md:116]: How can `platform` contain an arbitrary value?  If a Vendor is meant to interpret this information to be able to identify a user, this can not be arbitrary.

File: CCPA/Data Deletion Request Handling.md (hunk 9 above)
Comment (chrispaterson) [reply CCPA/Data Deletion Request Handling.md:219]: No it doesn't.  Param is the last value and it's optional, so no need to pass it unless it is defined.

File: CCPA/Data Deletion Request Handling.md [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+  </tr>
+</table>
+
+
+## Example Implementation
+
+Publisher site setup:
+
+```
+<html>
```
Comment (chrispaterson) [inline CCPA/Data Deletion Request Handling.md:169]: This whole section is WAY too much.  There is already a specification that defines how to set up the `__usapiLocator` frame and stub

File: CCPA/Data Deletion Request Handling.md (hunk 15 above)
Comment (Facens) [reply CCPA/Data Deletion Request Handling.md:92]: You may need to pass the deletion for multiple apps at the same time?

File: CCPA/Data Deletion Request Handling.md (hunk 17 above)
Comment (Facens) [reply CCPA/Data Deletion Request Handling.md:116]: We've decided to keep it open in order to support additional platforms that aren't listed (knowing that we can't possibly list them all)

File: CCPA/Data Deletion Request Handling.md (hunk 16 above)
Comment (lon-pilot-mpf) [reply CCPA/Data Deletion Request Handling.md:111]: app_identifier I believe @chrispaterson

File: CCPA/Data Deletion Request Handling.md (hunk 9 above)
Comment (lon-pilot-mpf) [reply CCPA/Data Deletion Request Handling.md:219]: It's already there I think. I may have missed it in my original comment or it may have been added (not honestly sure which). It's on line 229 below.

File: CCPA/Data Deletion Request Handling.md [hunk 19]
```diff
@@ -0,0 +1,330 @@
...
+	  }
+	}, false);
+
+
+	function vendorXDeletion() {
+	  //... do some deletion work …
+	  return;
+	}
+
+	__uspapi('registerDeletion', 1, vendorXDeletion);
```
Comment (lon-pilot-mpf) [inline CCPA/Data Deletion Request Handling.md]: I think this call to `__uspapi` needs to move outside the closing bracket on line 328 right @janwinkler?

File: CCPA/Data Deletion Request Handling.md (hunk 12 above)
Comment (lon-pilot-mpf) [reply CCPA/Data Deletion Request Handling.md:271]: Except for possibly my comment below it looks correct to me.

File: CCPA/Data Deletion Request Handling.md (hunk 17 above)
Comment (chrispaterson) [reply CCPA/Data Deletion Request Handling.md:116]: @Facens if it's arbitrary it's not usable because this information enables a Vendor script to interpret the other two values.  For example: If I'm a Vendor and I receive a `"platform"` value of `"foo-bar-23"` and it's not a platform I recognize, what do I do?

File: CCPA/Data Deletion Request Handling.md (hunk 15 above)
Comment (chrispaterson) [reply CCPA/Data Deletion Request Handling.md]: @Facens this software only runs in one App at a time
